### PR TITLE
Helper changes for symbolic testing

### PR DESCRIPTION
**Pull request by denis-bogdanas**

### CODE REVIEW

File: edsl.md [hunk 1]
```diff
@@ -54,9 +54,11 @@ where `F1 : F2 : F3 : F4` is the (two's complement) byte-array representation of
 
     syntax ByteArray ::= #abiCallData ( String , TypedArgs ) [function]
  // -------------------------------------------------------------------
-    rule #abiCallData( FNAME , ARGS )
-      => #parseByteStack(substrString(Keccak256(#generateSignature(FNAME, ARGS)), 0, 8))
-      ++ #encodeArgs(ARGS)
+    rule #abiCallData( FNAME , ARGS ) => #signatureCallData(FNAME, ARGS) ++ #encodeArgs(ARGS)
+
+    syntax ByteArray ::= #signatureCallData  ( String, TypedArgs ) [function]
+ // -------------------------------------------------------------------
```
Comment (ehildenb) [inline edsl.md:60]: formatting

File: edsl.md [hunk 2]
```diff
@@ -177,7 +179,7 @@ where `F1 : F2 : F3 : F4` is the (two's complement) byte-array representation of
 
     //Byte array buffer. Lemmas defined in evm-data-symbolic.k
     // SIZE, DATA // left zero padding
-    syntax ByteArray ::= #buf ( Int , Int ) [function, smtlib(buf)]
+    syntax ByteArray ::= #buf ( Int , Int ) [function, functional, smtlib(buf)]
  // ---------------------------------------------------------------
```
Comment (ehildenb) [inline edsl.md:183]: formatting

File: evm-types.md [hunk 3]
```diff
@@ -478,11 +479,11 @@ Most of EVM data is held in local memory.
...
     syntax Memory = Bytes
-    syntax Memory ::= Memory "[" Int ":=" ByteArray "]" [function, klabel(mapWriteBytes)]
- // -------------------------------------------------------------------------------------
+    syntax Memory ::= Memory "[" Int ":=" ByteArray "]" [function, functional, klabel(mapWriteBytes)]
+ // -------------------------------------------------------------------------------------------------
     rule WS [ START := WS' ] => replaceAtBytes(padRightBytes(WS, START +Int #sizeByteArray(WS'), 0), START, WS')  [concrete]
 
-    syntax ByteArray ::= #range ( Memory , Int , Int ) [function]
+    syntax ByteArray ::= #range ( Memory , Int , Int ) [function, functional]
  // -------------------------------------------------------------
```
Comment (ehildenb) [inline evm-types.md:487]: formatting

File: evm-types.md [hunk 4]
```diff
@@ -597,8 +598,8 @@ The local memory of execution is a byte-array (instead of a word-array).
  // ----------------------------------------------------------------------------------------------------------------
     rule #sizeByteArray ( WS ) => lengthBytes(WS) [concrete]
 
-    syntax ByteArray ::= #padToWidth      ( Int , ByteArray ) [function]
-                       | #padRightToWidth ( Int , ByteArray ) [function]
+    syntax ByteArray ::= #padToWidth      ( Int , ByteArray ) [function, functional]
+                       | #padRightToWidth ( Int , ByteArray ) [function, functional]
  // --------------------------------------------------------------------
```
Comment (ehildenb) [inline evm-types.md:603]: formatting

File: evm-types.md [hunk 5]
```diff
@@ -682,9 +683,9 @@ Addresses
 -   `#lookup` looks up a key in a map and returns 0 if the key doesn't exist, otherwise returning its value.
 
 ```k
-    syntax Int ::= #lookup ( Map , Int ) [function]
+    syntax Int ::= #lookup ( Map , Int ) [function, functional]
  // -----------------------------------------------
-    rule [#lookup.some]: #lookup( (KEY |-> VAL) M, KEY ) => VAL
+    rule [#lookup.some]: #lookup( (KEY |-> {VAL}:>Int) M, KEY ) => VAL
     rule [#lookup.none]: #lookup(               M, KEY ) => 0 requires notBool KEY in_keys(M)
```
Comment (ehildenb) [inline evm-types.md:689]: formatting

File: serialization.md [hunk 6]
```diff
@@ -21,7 +21,7 @@ Address/Hash Helpers
 -   `keccak` serves as a wrapper around the `Keccak256` in `KRYPTO`.
 
 ```k
-    syntax Int ::= keccak ( ByteArray ) [function, smtlib(smt_keccak)]
+    syntax Int ::= keccak ( ByteArray ) [function, functional, smtlib(smt_keccak)]
  // ------------------------------------------------------------------
```
Comment (ehildenb) [inline serialization.md:25]: formatting

File: edsl.md (hunk 2 above)
Comment (denis-bogdanas) [reply edsl.md:183]: what's wrong with formatting here?

File: evm-types.md (hunk 5 above)
Comment (denis-bogdanas) [reply evm-types.md:689]: ?

Comment (ehildenb) [reply evm-types.md]: Before the change the `=>` were lined up and the `-------` extended to the end of the previous line.

File: edsl.md (hunk 2 above)
Comment (ehildenb) [reply edsl.md:183]: `---` to end of previous line

File: edsl.md (hunk 1 above)
Comment (ehildenb) [reply edsl.md:60]: `---` to end of previous line

File: evm-types.md [hunk 7]
```diff
@@ -682,10 +683,10 @@ Addresses
 -   `#lookup` looks up a key in a map and returns 0 if the key doesn't exist, otherwise returning its value.
 
 ```k
-    syntax Int ::= #lookup ( Map , Int ) [function]
+    syntax Int ::= #lookup ( Map , Int ) [function, functional]
  // -----------------------------------------------
```
Comment (ehildenb) [inline evm-types.md:687]: formatting

File: evm-types.md [hunk 8]
```diff
@@ -682,10 +686,10 @@ Addresses
...
 
 ```k
-    syntax Int ::= #lookup ( Map , Int ) [function]
- // -----------------------------------------------
-    rule [#lookup.some]: #lookup( (KEY |-> VAL) M, KEY ) => VAL
-    rule [#lookup.none]: #lookup(               M, KEY ) => 0 requires notBool KEY in_keys(M)
+    syntax Int ::= #lookup ( Map , Int ) [function, functional]
+ // -----------------------------------------------------------
+    rule [#lookup.some]: #lookup( (KEY |-> {VAL}:>Int) M, KEY ) => VAL
+    rule [#lookup.none]: #lookup(                      M, KEY ) => 0 requires notBool KEY in_keys(M)
```
Comment (ehildenb) [inline evm-types.md:692]: We should make the second rule `owise`, to cover the case where `VAL` is not an `Int`, if we want to make this total.

File: tests/specs/lemmas.k [hunk 9]
```diff
@@ -5,6 +5,34 @@ module LEMMAS
...
+      requires N >Int M andBool N -Int M >=Int #sizeByteArray(BUF')                                               [simplification]
+
+    rule M [ N := BUF ] [ N := BUF' ] => M [ N := BUF' ] requires #sizeByteArray(BUF) ==Int #sizeByteArray(BUF')  [simplification]
+    //when BUF1 and BUF2 partially overlap and cannot be sorted by sorting rule.
+    rule MEM [ N := BUF1 ] [ M := BUF2:ByteArray ] [ N := BUF3 ] => MEM [ M := BUF2 ] [ N := BUF3 ]
+      requires #sizeByteArray(BUF1) ==Int #sizeByteArray(BUF3)                                                    [simplification]
+
+    rule BUF [ L .. W ] => .ByteArray requires W <=Int 0                    [simplification]
+    rule BUF [ 0 .. W ] => BUF        requires W ==Int #sizeByteArray(BUF)  [simplification]
+    rule BUF [ L .. W ] => .ByteArray requires L  >Int #sizeByteArray(BUF)  [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:22]: We can't add this third rule to both modules because the semantics are different for Java and Haskell implementation. Java implementation will return the empty `.ByteArray` like this says, and Haskell implementation will return the zero-padded bytearray of size `W`.

File: tests/specs/lemmas.k [hunk 10]
```diff
@@ -5,6 +5,34 @@ module LEMMAS
...
+
+    rule #sizeByteArray(BUF1 ++ BUF2)                  => #sizeByteArray(BUF1) +Int #sizeByteArray(BUF2)  [simplification]
+    rule #sizeByteArray(#buf(SIZE, _))                 => SIZE                                            [simplification]
+    rule #sizeByteArray(#buf(SIZE, _)[START .. WIDTH]) => WIDTH
+      requires #range(0 <= START < SIZE)
+       andBool #range(0 < WIDTH <= SIZE -Int START)                                                       [simplification]
+
+    rule #sizeByteArray(#range(_,_, N)) => N                                                              [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:30]: Please put this with the other `#sizeByteArray` lemmas.

File: tests/specs/lemmas.k [hunk 11]
```diff
@@ -5,6 +5,34 @@ module LEMMAS
...
+    rule .ByteArray ++ WS => WS  [simplification]
+
+    //Byte array store operations sorting. Slices with lower index first. Helps applying rule below.
+    rule MEM [ N := BUF:ByteArray ] [ M := BUF' ] => MEM [ M := BUF' ] [ N := BUF ]
+      requires N >Int M andBool N -Int M >=Int #sizeByteArray(BUF')                                               [simplification]
+
+    rule M [ N := BUF ] [ N := BUF' ] => M [ N := BUF' ] requires #sizeByteArray(BUF) ==Int #sizeByteArray(BUF')  [simplification]
+    //when BUF1 and BUF2 partially overlap and cannot be sorted by sorting rule.
+    rule MEM [ N := BUF1 ] [ M := BUF2:ByteArray ] [ N := BUF3 ] => MEM [ M := BUF2 ] [ N := BUF3 ]
+      requires #sizeByteArray(BUF1) ==Int #sizeByteArray(BUF3)                                                    [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:18]: This rule is too specific for my liking, let's not add it here. The reason is that what if there are 2 intermediate writes instead? Or 3? etc...

Let's remove it for now.

File: evm-types.md (hunk 8 above)
Comment (denis-bogdanas) [reply evm-types.md:693]: replaced with `#Ceil`

File: tests/specs/lemmas.k (hunk 9 above)
Comment (denis-bogdanas) [reply tests/specs/lemmas.k:22]: done. Separate lemmas for Java and Haskell

File: tests/specs/lemmas.k (hunk 11 above)
Comment (denis-bogdanas) [reply tests/specs/lemmas.k:18]: ok

**Comment by denis-bogdanas** [comment]
@ehildenb Except one question above this is ready for review. Should pass the tests.

File: tests/specs/lemmas.k (hunk 10 above)
Comment (ehildenb) [reply tests/specs/lemmas.k:26]: I think some code got removed in the meantime that moved these together.

File: tests/specs/lemmas.k [hunk 12]
```diff
@@ -213,17 +229,7 @@ module LEMMAS-HASKELL [symbolic, kore]
...
-
-    rule #sizeByteArray(BUF1 ++ BUF2)                  => #sizeByteArray(BUF1) +Int #sizeByteArray(BUF2)  [simplification]
-    rule #sizeByteArray(#buf(SIZE, _))                 => SIZE                                            [simplification]
-    rule #sizeByteArray(#buf(SIZE, _)[START .. WIDTH]) => WIDTH
-      requires #range(0 <= START < SIZE)
-       andBool #range(0 < WIDTH <= SIZE -Int START)                                                       [simplification]
-
-    rule BUF [ L .. W ] => .ByteArray requires W <=Int 0                    [simplification]
-    rule BUF [ 0 .. W ] => BUF        requires W ==Int #sizeByteArray(BUF)  [simplification]
+    rule BUF [ L .. W ] => padRightBytes(.Bytes, W, 0) requires L >=Int #sizeByteArray(BUF)  [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:232]: I think this rule can be implemented directly in the semantics instead, in `evm-types.md` around line 594.

```
    rule WS [ START .. WIDTH ] => substrBytes(padRightBytes(WS, START +Int WIDTH, 0), START, START +Int WIDTH) requires START <Int #sizeByteArray(WS) [concrete] 
    rule WS [ START .. WIDTH ] => padRightBytes(.Bytes, WIDTH, 0)                                              [owise, concrete] 
```

becomes (just removing `concrete` from the `owise` rule):

```
    rule WS [ START .. WIDTH ] => substrBytes(padRightBytes(WS, START +Int WIDTH, 0), START, START +Int WIDTH) requires START <Int #sizeByteArray(WS) [concrete] 
    rule WS [ START .. WIDTH ] => padRightBytes(.Bytes, WIDTH, 0)                                              [owise] 
```

I think this achieves the same effect, which is to say "if the prover can show that `notBool START <Int #sizeByteArray(WS)`, then simplify to the empty `.Bytes`".

File: tests/specs/lemmas.k [hunk 13]
```diff
@@ -5,20 +5,42 @@ module LEMMAS
...
 module LEMMAS-JAVA [symbolic, kast]
     imports EVM
     imports EDSL
     imports K-REFLECTION
 
-    rule #asWord( BUF => #drop(1, BUF) ) requires BUF [ 0 ] ==Int 0  [simplification]
+    rule BUF [ L .. W ] => .ByteArray requires L >=Int #sizeByteArray(BUF)  [simplification]
 
-    rule WS ++ .ByteArray => WS  [simplification]
+    rule #asWord( BUF => #drop(1, BUF) ) requires BUF [ 0 ] ==Int 0  [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:42]: @daejunpark are these lemmas sound for teh Java backend?

**Comment by ehildenb** [comment]
It looks pretty good to me, but can @daejunpark look and @virgil-serbanuta as well?

File: tests/specs/lemmas.k (hunk 12 above)
Comment (denis-bogdanas) [reply tests/specs/lemmas.k:232]: ok

File: tests/specs/lemmas.k (hunk 13 above)
Comment (denis-bogdanas) [reply tests/specs/lemmas.k:42]: 2nd lemma was there since before my changes. 1st is probably not needed for Java proofs, but looks valid.

File: evm-types.md [hunk 14]
```diff
@@ -591,7 +598,7 @@ The local memory of execution is a byte-array (instead of a word-array).
     syntax ByteArray ::= ByteArray "[" Int ".." Int "]" [function]
  // --------------------------------------------------------------
     rule WS [ START .. WIDTH ] => substrBytes(padRightBytes(WS, START +Int WIDTH, 0), START, START +Int WIDTH) requires START <Int #sizeByteArray(WS) [concrete]
-    rule WS [ START .. WIDTH ] => padRightBytes(.Bytes, WIDTH, 0)                                              [owise, concrete]
+    rule WS [ START .. WIDTH ] => padRightBytes(.Bytes, WIDTH, 0)                                              [owise]
```
Comment (virgil-serbanuta) [inline evm-types.md:601]: I'm curious: Why did you remove `concrete` here, but not for the rule above?

Comment (ehildenb) [reply evm-types.md:601]: Removing `concrete` allows it to simplify in more cases to the empty bytearray, otherwise there were specific lemmas added which simplified in cases that it could. This is just more general than all those lemmas were.

File: edsl.md [hunk 15]
```diff
@@ -182,6 +184,7 @@ where `F1 : F2 : F3 : F4` is the (two's complement) byte-array representation of
     rule #buf(SIZE, DATA) => #padToWidth(SIZE, #asByteStack(DATA))
       requires #range(0 <= DATA < (2 ^Int (SIZE *Int 8)))
       [concrete]
+    rule #Ceil(#buf(SIZE, _)) => {(0 <=Int SIZE) #Equals true}  [anywhere]
```
[the remark this answers is not in the document]
Comment (denis-bogdanas) [reply edsl.md:187]: Yes you are right. It might need additional lemmas though. Will have to fix this.

File: evm-types.md [hunk 16]
```diff
@@ -423,11 +423,18 @@ A cons-list is used for the EVM wordstack.
...
 ```
 
--   Definedness conditions for `WS [ N ]` and `WS [ N := W ]`
+-   Definedness conditions for `WS [ N ]`, `WS [ N := W ]`, `BA [ N .. W ]`.
 
 ```{.k .symbolic}
-    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
-    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
+    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
```
[the remark this answers is not in the document]
Comment (denis-bogdanas) [reply evm-types.md:430]: I added it here:
```
rule #Ceil(BA[ N := _:ByteArray ]) => {(0 <=Int N)                                        #Equals true}  [anywhere]
```

**Comment by denis-bogdanas** [comment]
@ehildenb @daejunpark All comments addressed. I see it passes now. No idea why it was failing before. Changes are unrelated to failure. Ready for re-review.

File: evm-types.md [hunk 17]
```diff
@@ -423,11 +423,18 @@ A cons-list is used for the EVM wordstack.
...
+    rule #Ceil(BA[ N .. W ]) => {((0 <=Int N) andBool (0 <=Int W))                            #Equals true}  [anywhere]
+    rule #Ceil(BA[ N := _:ByteArray ]) => {(0 <=Int N)                                        #Equals true}  [anywhere]
+    rule #Ceil( #range(LM, START, WIDTH) ) => {((0 <=Int START) andBool (0 <=Int WIDTH))      #Equals true}  [anywhere]
+    rule #Ceil(#padToWidth(N, _))      => {(0 <=Int N)                                        #Equals true}  [anywhere]
```
Comment (virgil-serbanuta) [inline evm-types.md:434]: I'm not sure this is correct.

I think that `#padToWidth` is, basically, defined as `padLeftBytes` (at least when working with `ByteArray`), and, in `domains.k`, `padLeftBytes` has two definitions: as a builtin and as a normal function. Interestingly enough, when it is defined as a normal function, it seems to be a total function. So, in the past, I wanted to mark it as a total function, but I think that @ehildenb explained to me that, being a builtin, whoever implements it can choose whether it is total or not. I hope I'm remembering this correctly.

Anyway, if that's indeed the case, then you can't define this `ceil` as you did since it may be different for different backends. Possible solutions include
```
#Ceil(#padToWidth(...))      => #Ceil(padLeftBytes(...))
```
Hopefully the Haskell backend would be able to solve that at some future point. Redefining `#padToWidth` to make it explicit that it's undefined when `N<0` would also work, I guess.

File: evm-types.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (daejunpark) [reply evm-types.md:430]: No, that’s not. Please see carefully the syntax. The last argument is of Int not ByteArray. This shows again how confusing the abuse of notation is!

File: evm-types.md (hunk 17 above)
Comment (denis-bogdanas) [reply evm-types.md:434]: I guess we need this rule now. Eventually will add a todo. What do you mean "Redefining #padToWidth to make it explicit that it's undefined when N<0 would also work" ? Isn't `#Ceil` the explicit way to define when function is undefined?

File: evm-types.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (denis-bogdanas) [reply evm-types.md:430]: Thanks. I guess I'm still a bit dizzy.

File: evm-types.md (hunk 17 above)
Comment (virgil-serbanuta) [reply evm-types.md:434]: Redefining `#padToWidth`:
```
    rule #padToWidth(N, BS)      => padLeftBytes(BS, N, 0) requires N >=Int 0  [concrete]
```

The main question is whether `#Ceil` should be in sync with the function definition rules and the builtin implementation. If the `ceil` axiom says that the function is defined iff `n>=0`, but the backend believes that the same function is total, is that reasonable? If `phi` is a formula, then `phi` should be equivalent to `phi and ceil(phi)`. So we could evaluate `#padToWidth` in two inconsistent ways:

```
#padToWidth(-1, B) = B  // evaluated as a builtin.

#padToWidth(-1, B) = #padToWidth(-1, B) and ceil(#padToWidth(-1, B)) = #padToWidth(-1, B) and bottom = bottom.
```

@traiansf @ttuegel in case they have comments.

Comment (denis-bogdanas) [reply evm-types.md]: @virgil-serbanuta I incorporated your suggestions but also added these 2 #Ceil lemmas:
```
    rule #Ceil(padLeftBytes(_, N, _))       => {(0 <=Int N)                                   #Equals true}  [anywhere]
    rule #Ceil(padRightBytes(_, N, _))      => {(0 <=Int N)                                   #Equals true}  [anywhere]
```
We still need to resolve `#Ceil` formulas for these functions. Alternative would be to add them to `domains.k`

**Comment by ehildenb** [comment]
Please don't merge the update submodule PR into this one @denis-bogdanas . Instead, you can just approve that PR and get it merged directly, then merge master into this one.

If other PRs update the submodules, then the submodule updater gets confused, it can't handle merge conflicts at all.

**Comment by denis-bogdanas** [comment]
This PR would pass tests with latest K, but K update PR fails for other reasons as you can see. That one must be fixed first.

**Comment by denis-bogdanas** [comment]
@ehildenb We need updated K for this PR to pass. Other than that it looks ready.

**Comment by ehildenb** [comment]
We're waiting on a regression in the LLVM backend to update K. I'll keep you posted, thanks!

File: evm-types.md [hunk 18]
```diff
@@ -423,11 +423,20 @@ A cons-list is used for the EVM wordstack.
...
-
-```{.k .symbolic}
-    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
-    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
+-   Definedness conditions for `WS [ N ]`, `WS [ N := W ]`, `BA [ N .. W ]`.
+
+```{.k .symbolic-bytes}
+    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(BA[ N .. W ]) => {((0 <=Int N) andBool (0 <=Int W))                            #Equals true}  [anywhere]
```
Comment (ehildenb) [inline evm-types.md:431]: I think we don't need it anymore @denis-bogdanas ?

File: evm-types.md [hunk 19]
```diff
@@ -423,11 +423,20 @@ A cons-list is used for the EVM wordstack.
...
+    rule #Ceil(BA[ N := _:ByteArray ]) => {(0 <=Int N)                                        #Equals true}  [anywhere]
+    rule #Ceil( #range(LM, START, WIDTH) ) => {((0 <=Int START) andBool (0 <=Int WIDTH))      #Equals true}  [anywhere]
```
Comment (ehildenb) [inline evm-types.md:433]: Same here, do we need it anymore?

**Comment by denis-bogdanas** [comment]
@ehildenb I re-rebased and removed `#Ceil` rules that are not needed. Only the rules for `[..]` and `#range` were removed. For the rest, functions are not `functional` so `#Ceil` is still needed.

File: tests/specs/lemmas.k [hunk 20]
```diff
@@ -5,20 +5,42 @@ module LEMMAS
...
+    rule WS ++ .ByteArray => WS  [simplification]
+    rule .ByteArray ++ WS => WS  [simplification]
+
+    //Byte array store operations sorting. Slices with lower index first. Helps applying rule below.
+    rule MEM [ N := BUF:ByteArray ] [ M := BUF' ] => MEM [ M := BUF' ] [ N := BUF ]
+      requires N >Int M andBool N -Int M >=Int #sizeByteArray(BUF')                                               [simplification]
+
+    rule M [ N := BUF ] [ N := BUF' ] => M [ N := BUF' ] requires #sizeByteArray(BUF) ==Int #sizeByteArray(BUF')  [simplification]
+
+    rule BUF [ L .. W ] => .ByteArray requires W <=Int 0                    [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:17]: Is this one needed anymore with the new rules for `_[_.._]`?

Comment (denis-bogdanas) [reply tests/specs/lemmas.k]: It is still needed for case `W ==Int 0`. I simplified the rule.

File: tests/specs/lemmas.k [hunk 21]
```diff
@@ -5,20 +5,42 @@ module LEMMAS
...
+    rule BUF [ L .. W ] => .ByteArray requires W ==Int 0                    [simplification]
+    rule BUF [ 0 .. W ] => BUF        requires W ==Int #sizeByteArray(BUF)  [simplification]
+
+    rule #sizeByteArray(BUF1 ++ BUF2)                  => #sizeByteArray(BUF1) +Int #sizeByteArray(BUF2)  [simplification]
+    rule #sizeByteArray(#buf(SIZE, _))                 => SIZE                                            [simplification]
+    rule #sizeByteArray(#buf(SIZE, _)[START .. WIDTH]) => WIDTH
+      requires #range(0 <= START < SIZE)
+       andBool #range(0 < WIDTH <= SIZE -Int START)                                                       [simplification]
+
+    rule #sizeByteArray(#range(_,_, N)) => N                                                              [simplification]
```
Comment (ehildenb) [inline tests/specs/lemmas.k:26]: Should this have a side-condition that `N >=Int 0` and `START >=Int 0`?

Comment (ehildenb) [reply tests/specs/lemmas.k:26]: Because if we have `#sizeByteArray(BUF) == 10`:

we have two different reductions:

with the above rule:

```
#sizeByteArray(#range(BUF, -10, 5)) => 5
```

and:

```
#sizeByteArray(#range(BUF, 10, -5)) => #sizeByteArray(.ByteArray) => 0
```

Comment (denis-bogdanas) [reply tests/specs/lemmas.k]: Corrected

File: evm-types.md [hunk 22]
```diff
@@ -425,9 +425,16 @@ A cons-list is used for the EVM wordstack.
 
 -   Definedness conditions for `WS [ N ]` and `WS [ N := W ]`
 
-```{.k .symbolic}
-    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
-    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
+```{.k .symbolic-bytes}
+    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
```
Comment (ehildenb) [inline evm-types.md:430]: Can't we just make `_[_]` and `_[_:=_]` total instead of adding definedness conditions? I think it should be possible pretty simply by adding cases:

```
rule .WordStack [ 0 ] => 0

rule .WordStack [ N := WS ] => #take(N, .WordStack) ++ WS
```

Then we don't need to maintain definedness conditions.

File: edsl.md [hunk 23]
```diff
@@ -182,6 +184,9 @@ where `F1 : F2 : F3 : F4` is the (two's complement) byte-array representation of
     rule #buf(SIZE, DATA) => #padToWidth(SIZE, #asByteStack(DATA))
       requires #range(0 <= DATA < (2 ^Int (SIZE *Int 8)))
       [concrete]
+    //todo prohibitive performance drop for ERC20 verifier when full rule is used.
+    //rule #Ceil(#buf(SIZE, DATA)) => {(0 <=Int SIZE) andBool #range(0 <= DATA < (2 ^Int (SIZE *Int 8))) #Equals true}  [anywhere]
+    rule #Ceil(#buf(SIZE, DATA)) => {(0 <=Int SIZE) #Equals true}  [anywhere]
```
Comment (ehildenb) [inline edsl.md:189]: I would rather not assert false things here if possible.

File: evm-types.md [hunk 24]
```diff
@@ -425,9 +425,16 @@ A cons-list is used for the EVM wordstack.
...
+    rule #Ceil(BA[ N := _:ByteArray ]) => {(0 <=Int N)                                        #Equals true}  [anywhere]
+    rule #Ceil(#padToWidth(N, BS))          => #Ceil(padLeftBytes(BS, N, 0))                                 [anywhere]
+    rule #Ceil(padLeftBytes(_, N, _))       => {(0 <=Int N)                                   #Equals true}  [anywhere]
+    rule #Ceil(#padRightToWidth(N, BS))     => #Ceil(padRightBytes(BS, N, 0))                                [anywhere]
+    rule #Ceil(padRightBytes(_, N, _))      => {(0 <=Int N)                                   #Equals true}  [anywhere]
```
Comment (ehildenb) [inline evm-types.md:435]: These should be added to `domains.k`, not here.

File: evm-types.md [hunk 25]
```diff
@@ -425,9 +425,16 @@ A cons-list is used for the EVM wordstack.
 
 -   Definedness conditions for `WS [ N ]` and `WS [ N := W ]`
 
-```{.k .symbolic}
-    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
-    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
+```{.k .symbolic-bytes}
+    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(BA[ N := _:ByteArray ]) => {(0 <=Int N)                                        #Equals true}  [anywhere]
```
Comment (ehildenb) [inline evm-types.md:431]: Here as well, I think we can make `_[_:=_:ByteArray]` total instead.

File: edsl.md (hunk 23 above)
Comment (denis-bogdanas) [reply edsl.md:189]: What's your solution then?

Comment (ehildenb) [reply edsl.md:189]: I don't have one, but I don't really want to merge this into KEVM, because then other proofs may depend on it. How big is the slowdown?

Maybe if this isn't fixable, then we should just not merge this now, and consider it for the next "controversial changes" PR.

File: evm-types.md [hunk 26]
```diff
@@ -412,22 +412,27 @@ A cons-list is used for the EVM wordstack.
 -   `WS [ N := W ]` sets element $N$ of $WS$ to $W$ (padding with zeros as needed).
 
 ```k
-    syntax Int ::= WordStack "[" Int "]" [function]
+    syntax Int ::= WordStack "[" Int "]" [function, functional]
  // -----------------------------------------------
```
Comment (ehildenb) [inline evm-types.md:416]: formatting

File: evm-types.md [hunk 27]
```diff
@@ -412,22 +412,27 @@ A cons-list is used for the EVM wordstack.
...
     rule (W : _):WordStack [ N ] => W                  requires N ==Int 0
     rule WS:WordStack      [ N ] => #drop(N, WS) [ 0 ] requires N  >Int 0
+    rule WS                [ N ] => 0                  requires N  <Int 0
 
-    syntax WordStack ::= WordStack "[" Int ":=" Int "]" [function]
+    syntax WordStack ::= WordStack "[" Int ":=" Int "]" [function, functional]
  // --------------------------------------------------------------
```
Comment (ehildenb) [inline evm-types.md:422]: formatting

File: evm-types.md [hunk 28]
```diff
@@ -478,9 +483,10 @@ Most of EVM data is held in local memory.
 
 ```{.k .membytes}
     syntax Memory = Bytes
-    syntax Memory ::= Memory "[" Int ":=" ByteArray "]" [function, klabel(mapWriteBytes)]
+    syntax Memory ::= Memory "[" Int ":=" ByteArray "]" [function, functional, klabel(mapWriteBytes)]
  // -------------------------------------------------------------------------------------
```
Comment (ehildenb) [inline evm-types.md:487]: formatting

**Comment by denis-bogdanas** [comment]
@ehildenb ready for review

File: evm-types.md [hunk 29]
```diff
@@ -412,22 +412,27 @@ A cons-list is used for the EVM wordstack.
...
     rule WS:WordStack      [ N ] => #drop(N, WS) [ 0 ] requires N  >Int 0
+    rule WS:WordStack      [ N ] => 0                  requires N  <Int 0
 
-    syntax WordStack ::= WordStack "[" Int ":=" Int "]" [function]
- // --------------------------------------------------------------
+    syntax WordStack ::= WordStack "[" Int ":=" Int "]" [function, functional]
+ // --------------------------------------------------------------------------
     rule (W0 : WS):WordStack [ N := W ] => W  : WS                     requires N ==Int 0
     rule (W0 : WS):WordStack [ N := W ] => W0 : (WS [ N -Int 1 := W ]) requires N  >Int 0
+    rule        WS:WordStack [ N := W ] => .WordStack                  requires N  <Int 0
```
Comment (ehildenb) [inline evm-types.md:425]:
```suggestion
    rule        WS:WordStack [ N := W ] => WS                                requires N  <Int 0
```

File: evm.md [hunk 30]
```diff
@@ -2496,7 +2496,7 @@ After interpreting the strings representing programs as a `WordStack`, it should
 -   `#dasmOpCode` interperets a `Int` as an `OpCode`.
 
 ```k
-    syntax OpCode ::= #dasmOpCode ( Int , Schedule ) [function]
+    syntax OpCode ::= #dasmOpCode ( Int , Schedule ) [function, memo]
  // -----------------------------------------------------------
```
Comment (ehildenb) [inline evm.md:2500]:
```suggestion
 // -----------------------------------------------------------------
```

File: evm-types.md [hunk 31]
```diff
@@ -602,8 +608,8 @@ The local memory of execution is a byte-array (instead of a word-array).
     syntax ByteArray ::= #padToWidth      ( Int , ByteArray ) [function]
                        | #padRightToWidth ( Int , ByteArray ) [function]
  // --------------------------------------------------------------------
-    rule #padToWidth(N, BS)      => padLeftBytes(BS, N, 0)  [concrete]
-    rule #padRightToWidth(N, BS) => padRightBytes(BS, N, 0) [concrete]
+    rule #padToWidth(N, BS)      =>  padLeftBytes(BS, N, 0) requires N >=Int 0  [concrete]
+    rule #padRightToWidth(N, BS) => padRightBytes(BS, N, 0) requires N >=Int 0  [concrete]
```
Comment (ehildenb) [inline evm-types.md:612]:
```suggestion
    rule #padToWidth(N, BS)      =>  padLeftBytes(BS, N, 0) requires          N >=Int 0  [concrete]
    rule #padToWidth(N, BS)      =>  BS                     requires notBool (N >=Int 0)
    rule #padRightToWidth(N, BS) => padRightBytes(BS, N, 0) requires          N >=Int 0  [concrete]
    rule #padRightToWidth(N, BS) => BS                      requires notBool (N >=Int 0)
```
